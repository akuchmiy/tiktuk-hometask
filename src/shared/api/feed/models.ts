export interface AuthorMeta {
  id: string
  name: string
  avatar: string
  heart: number
  video: number
  fans: number
  following: number
}

export interface Hashtag {
  id: string
  name: string
}

export interface Feed {
  id: string
  text: string
  createTime: number
  authorMeta: AuthorMeta
  videoUrl: string
  videoMeta: {
    height: number
    width: number
  }
  hashtags: Hashtag[]
  playCount: number
  diggCount: number
  commentCount: number
}
