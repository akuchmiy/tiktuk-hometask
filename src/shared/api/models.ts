export * from './feed/models'
export * from './user/models'
