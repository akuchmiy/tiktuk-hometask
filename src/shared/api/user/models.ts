export interface UserData {
  user: User
  stats: UserStats
}

export interface UserStats {
  followerCount: number
  followingCount: number
  heart: number
  videoCount: number
  diggCount: number
}

export interface User {
  id: string
  uniqueId: string
  nickname: string
  avatarMedium: string
  avatarThumb: string
  signature: string
  bioLink: {
    link: string
    risk: number
  }
  verified: boolean
}
