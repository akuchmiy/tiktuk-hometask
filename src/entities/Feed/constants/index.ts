export const IS_INITIAL_PLAYING = false

export const OBSERVE_WHEN_COLUMNS = [1]

export const OBSERVER_OPTIONS = {
  threshold: 0.5,
}
