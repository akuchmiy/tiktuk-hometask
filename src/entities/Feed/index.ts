export * from './ui'

export * as feedModel from './model'
