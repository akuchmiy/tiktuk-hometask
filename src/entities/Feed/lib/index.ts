import React from 'react'

export const playVideo = async (video: HTMLVideoElement) => {
  try {
    await video.play()
    return true
  } catch (e) {
    return false
  }
}

export const pauseVideo = (video: HTMLVideoElement) => {
  video.pause()

  return false
}

export const toggleVideo = async (
  video: HTMLVideoElement
): Promise<boolean> => {
  if (!video.paused) return pauseVideo(video)

  return playVideo(video)
}

export const pauseVideos = (
  videos: React.MutableRefObject<HTMLVideoElement[]>
) => {
  videos.current.forEach((video) => pauseVideo(video))
}

export const scrollToVideo = (video: HTMLVideoElement) => {
  const rect = video.getBoundingClientRect()
  window.scrollTo({
    top: rect.top + window.scrollY - rect.height / 5,
    left: 0,
    behavior: 'smooth',
  })
}

export const observeVideos = (
  videos: React.MutableRefObject<HTMLVideoElement[]>,
  onVideoIntersect: IntersectionObserverCallback
) => {
  const observer = new IntersectionObserver(onVideoIntersect, {
    threshold: 0.5,
  })

  videos.current.forEach((video) => {
    if (video instanceof HTMLVideoElement) {
      observer.observe(video)
    }
  })

  return observer
}

export const trimHashtags = (text: string) => {
  const firstHashIndex = text.indexOf('#')
  if (firstHashIndex === -1) return text

  return text.slice(0, firstHashIndex)
}
