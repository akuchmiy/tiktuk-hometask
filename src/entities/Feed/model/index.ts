export * from './useFeed'
export * from './useVideoControls'
