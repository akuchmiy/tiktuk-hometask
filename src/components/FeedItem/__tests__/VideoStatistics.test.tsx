import { shallow } from 'enzyme'
import VideoStatistics from '../VideoStatistics'
import { Feed } from '../../../models/Feed'

jest.mock('../../../services/NumberService', () => ({
  formatNumber: (input: number) => input,
}))

const feed: Feed = {
  diggCount: 50,
  commentCount: 100,
} as Feed

describe('VideoStatistics tests', function () {
  it('should match snapshot', function () {
    const wrapper = shallow(<VideoStatistics feed={feed} />)

    expect(wrapper.debug()).toMatchSnapshot()
  })
})
