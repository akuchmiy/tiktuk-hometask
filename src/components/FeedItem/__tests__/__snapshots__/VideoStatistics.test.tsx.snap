// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoStatistics tests should match snapshot 1`] = `
"<div className=\\"flex flex-col gap-y-3.5 text-black dark:text-gray-100 leading-5 \\">
  <button aria-label=\\"Likes\\" className=\\"relative w-16 h-16 bg-gray-100 dark:bg-gray-600 rounded-full border-l-2 border-t-2 border-black dark:border-gray-100 transition-colors duration-100 hover:opacity-80\\">
    <FontAwesomeIcon className=\\"text-md\\" icon={{...}} border={false} mask={{...}} fixedWidth={false} inverse={false} flip={{...}} listItem={false} pull={{...}} pulse={false} rotation={{...}} size={{...}} spin={false} symbol={false} title=\\"\\" transform={{...}} swapOpacity={false} />
    <br />
    <span>
      50
    </span>
  </button>
  <button aria-label=\\"Comments\\" className=\\"relative w-16 h-16 bg-gray-100 dark:bg-gray-600 rounded-full border-l-2 border-t-2 border-black dark:border-gray-100 transition-colors duration-100 hover:opacity-80\\">
    <FontAwesomeIcon className=\\"text-md\\" icon={{...}} border={false} mask={{...}} fixedWidth={false} inverse={false} flip={{...}} listItem={false} pull={{...}} pulse={false} rotation={{...}} size={{...}} spin={false} symbol={false} title=\\"\\" transform={{...}} swapOpacity={false} />
    <br />
    <span>
      100
    </span>
  </button>
</div>"
`;
